### PR TITLE
Improvements for Manual Reprocessing

### PR DESCRIPTION
**Pull request by pdeperio**
- Allow specification of pax version and output directory for ROOT file
  - Can specify multiple versions and corresponding versions if desired, e.g. "head" processing and stable release "version" processing
- Also checks repository "hash" for the given pax version or head and includes in DB entry (redundant with "version" if head is enforced to be always at a stable release)
- Merge branch "[issue4](https://github.com/XENON1T/cax/tree/issue4)" for specifying custom config .json file. This fixes issue https://github.com/XENON1T/cax/issues/4
- Implement dataset list specification in config file as, "dataset_list": ["160512_1339", "160514_0827", ...]
- Implement task specification in config file as, "task_list": ["ProcessBatchQueue", ...]
- Implement Stockholm switch for submission script. This fixes issue https://github.com/XENON1T/cax/issues/12 after Boris confirms and tests.
  - Note, creates a new "cax-process" entry_point in setup.py

WARNING: The databasing during the actual job ("process" function) is still buggy with creating multiple entries when retrying, seeming to overwrite existing entries with new version, and maybe not retrying when we want it too. I'm not yet familiar enough with MongoDB to improve. @tunnell, since you've been cleaning up everything, can you please take a look?

### CODE REVIEW

File: cax/cax.json [hunk 1]
```diff
@@ -34,7 +34,9 @@
     "directory": "/project/lgrandi/xenon1t",
     "upload_options": [],
     "username": "tunnell",
-    "download_options": []
+    "download_options": [],
+    "pax_versions": ["v4.10.0"],
+    "pax_locations": ["/project/lgrandi/xenon1t/root/pax_v4.10.0"]
```
Comment (tunnell) [inline cax/cax.json]: Can this be a list of dicts?  "pax" : {"v4.10.0" : "/project/lgrandi/xenon1t/root/pax_v4.10.0"}

File: cax/config.py [hunk 2]
```diff
@@ -8,6 +8,8 @@
 
 import pymongo
 
+# global variable to store the specified .json config file
+cax_configure = ''
```
Comment (tunnell) [inline cax/config.py]: Make caps.  CAX_CONFIGURE.  Makes clear it's a global.

File: cax/config.py [hunk 3]
```diff
@@ -29,16 +31,33 @@ def get_hostname():
...
+    """Set the cax.json file at your own
+    """
+    global cax_configure
+    cax_configure = confg
+
+def load():
+
+    # User-specified config file
+    if cax_configure:
+        filename = os.path.abspath( cax_configure )
```
Comment (tunnell) [inline cax/config.py]: Add WARNING here to tell people custom cax.json?

File: cax/main.py [hunk 4]
```diff
@@ -14,12 +16,25 @@ def main():
...
     # Check passwords and API keysspecified
     mongo_password()
 
+    # Get specified cax.json configuration file for cax:
+    caxjson_config = args.config_file
+    if caxjson_config:
+        if not os.path.isfile( caxjson_config ):
+            logging.error("Config file %s not found", caxjson_config)
+        else:
+            #logging.info("Using custom config file: %s", caxjson_config) # seems to kill rest of output...
```
Comment (tunnell) [inline cax/main.py]: ?

File: cax/config.py (hunk 3 above)
Comment (pdeperio) [inline cax/config.py]: but this gets called every iteration and task i think...

File: cax/task.py [hunk 5]
```diff
@@ -15,10 +15,14 @@ def __init__(self):
...
         """Run this periodically"""
 
+        # Get user-specified list of datasets
+        datasets = config.get_dataset_list()
+
         # Collect all run documents.  This has to be turned into a list
         # to avoid timeouts if a task takes too long.
-        docs = list(self.collection.find({'detector': 'tpc',
-                                          'number' : {"$gt": 489}}))
+        for self.run_doc in self.collection.find({'detector': 'tpc'}):
```
Comment (tunnell) [inline cax/task.py]: I don't think you want this loop?

File: cax/cax.json (hunk 1 above)
Comment (pdeperio) [inline cax/cax.json]: ya, can you implement this quickly? (sorry i'm a noob)

File: cax/main.py (hunk 4 above)
Comment (pdeperio) [inline cax/main.py]: this was trying to warn people of using custom config (as per your previous comment), but for some reason it stopped all the other output info from printing to screen, i'm not sure why

**Comment by pdeperio** [comment]
I started to field test this on Midway: /project/lgrandi/deployHQ/cax_pax_process_version
installed under pax_v4.10.0 environment.

I started a daemon but then realized on the second iteration that some jobs were still queued and got submitted again. I've deleted those duplicate jobs and stopped the daemon for now, so @tunnell should check that queue is empty before (merging this and) restarting the daemon.

I've also removed most of the erroneous processed data DB entries that arose during my testing.

**Comment by pdeperio** [comment]
I think the previous commit fixes the queue issue. But now I'm seeing some strange behavior from CompareChecksums and RetryStalledTransfer:
```
/project/lgrandi/deployHQ/cax_pax_process_version/cax.log
```
starting around 2016-05-20 23:32:51

Some processed files being deleted, which I thought were OK, but maybe cax getting confused with existing copy of v4.10.0 processed with head?

**Comment by tunnell** [comment]
Yeah, the checksum for processed data won't work.  I just need to more and more generalize this.  At first, just raw.  Then processed.  Now many versions processed.

**Pull request merged**